### DISCUSSION TITLE
Lesson 12 couldn't  understand how "endingUserHealthFactor" is calculating

### DISCUSSION THREAD

**Discussion by miladad8** [Cyfrin/foundry-full-course-cu#2361] [Q&A]
Hi, I am having trouble understanding how `liquidate` function in `DSCEngine` is calculating `endingUserHealthFactor`. I tried to run `testMustImproveHealthFactorOnLiquidation` test and although it is passing, But it is hard to understand the logic behind it. I tried `console.log` to log the healthFactor values before and after liquidation. `startingUserHealthFactor` looks right to me. However the log value for `endingUserHealthFactor` equals zero and I wonder how it calculates it. Just let me know If you want my test code. However I would appreciate if someone explain how to calculate `endingUserHealthFactor` with an example.

**Comment by EngrPips** [comment]
Hello @miladad8, isn't the codebase supposed to calculate the `helathFactor` the same way irrespective of whether it is calculating `startingUserHealthFactor` or `endingUserHealthFactor`?

**Comment by miladad8** [comment]
Not quite the same I think. As far as I know, calculating `startingUserHealthFactor ` is very straightforward. However for calculating the user `healthFactor `after liquidation, we need to consider `debtTocover `and the amount of collateral taken away from the user, If I understand it well. That is actually where I get confused. Do you know any example in the repo that explains it? thank you

**Comment by EngrPips** [comment]
It seems like I didn't quite understand the codebase, but my little understanding was that we are always using the amount of `DSC` minted and the amount of `collateral` deposited to calculate the `health factor` of user whether it is a startingHealthFactor` or it is an `endingHealthFactor` nothing changes except for when you liquidate a user the amount of their `DSC` minted would have reduced because you liquidated them and their amount of `collateral` deposited would have reduces as well in a manner that makes them healthy.

**Comment by EngrPips** [comment]
Let's walk through a simple example together 

**AT THE MOMENT OF MINTING DSC**
UserA deposited collateral worth $100 to the protocol and minted `DSC` worth $40, if we calculate the health factor it will be as below 
userAHealthFactor = 50/40 = 1.25
Obviously, userA is healthy at this point in time 

**SOME MOMENT LATER**
userA collateral is now worth $70 because the value of their collateral dropped and they are still holding `dsc` worth $40, now their healthFactor will be the below 

userAHealthFactor = 35/40 = 0.875

at this point in time, userA is unhealthy, and we want a liquidator to liquidate them 

**THE MOMENT A LIQUIDATOR LIQUIDATES USERA**
At this moment a liquidator comes to liquidate userA and paid $30 out of the $40 that userA minted as debt now the maths will go as below 

userA mintedDscValue = $10 [This is because the liquidator have paid $30 out of the DSC they minted]
userA depositedCollateral = $70 - $33 = $37 [this because the liquidator will get extra 10% of what the liquidated]

liquidator USDValue = $33 [ the liquidator basically paid $30 to get $33, let's assume this is a profitable to the liquidator] 

The ending userA healthFactor will now be the below
endingUserA healthFactor = $18.5 / $10 = 1.85

Obviously, the liquidation process moves the healthFactor of userA from an unhealthy state to a healthy state and hence the process wouldn't revert.

If you notice you will see that regardless of whether we are calculating `startinghealthFactor` or `endingHealthFctor` the formula we use is the same, Yeah value changes but we do the calculation the exact same way. 

Please ask questions if anything is confusing.

**Reply by miladad8** [reply]
thank you so much for your time. I think the reason I get confused is because of the math and units($ and ETH). Could you please do the calculation for below setup and check if  I am doing it right? Thanks in advance.
USER deposited = 10 ether
USER minted = 100 ether
LIQUIDATOR debtToCover = 10 ether
depositedCollateral current value = 18 $

StartingUserHealthFactor =>  (10 * 18)/2=90 =>  90/100 = 0.9
when I console.log `StartingUserHealthFactor ` I got 0.9 value as well.
However for `endingUserHealthFactor ` console.log shows 0.
Actually I don't get `endingUserHealthFactor ` part.

**Reply by EngrPips** [reply]
How come a user deposited 10 ether worth of collateral and they are able to mint 100 ether worth of DSC, or am I reading this wrongly.

**Reply by miladad8** [reply]
Well this is the setup used in `testMustImproveHealthFactorOnLiquidation` test mentioned in repo associated with lesson 12. As far as I know maybe it's 100$ and the keyword `ether` here is just for decimal and precession stuff. So math and calculation in DSEngine functions would add up. It's just my guess :)
The test is passing However I was curious how it calculates the `endingUserHealthFactor ` as zero.

**Comment by EngrPips** [comment]
Honestly, I will say go back to the `DSCEngine` contract and study the `liquidation` process flow and process appropriately so you can understand how the calculations work precisely.

**Reply by miladad8** [reply]
Hi again! I go through the `liquidation` function to understand the logic. I think the key part is in ` getTokenAmountFromUsd` function. The argument that this function takes is `usdAmountInWei`  It's obviously saying that the $ value should be considered with additional 18 decimals. 
And here's the exact example that Patrick explains in the course:
`usdAmountInWei * PRECISION) / (uint256(price) * ADDITIONAL_FEED_PRECISION`
`10e18 * 1e18 / 18e18 `
So I think  `amountMinted = 100 ether` means 100$  worth of DSC with 18  decimals(I forgot to add $ in above).
However I still don't understand how `endingUserHealthFactor ` is calculated as zero in the scenario I mentioned above.
console.log shows the `endingUserHealthFactor ` as zero.

Here's how I calculate the `endingUserHealthFactor` :

Collateral current value = 18 $
USER deposited = 10 ether => considering collateral value as 18$ => 180$ e18
USER minted = 100$ ether = 100$ e18
LIQUIDATOR debtToCover = 10$ ether =10e18
Liquidator gets: 10$ e18+ 1$ e18 =11$ e18
USER Collateral after liquidation: 180$ e18 - 11$ e18 = 169$ e18
USER minted: 100$ e18 - 10$ e18 =90$ e18 
endingUserHealthFactor: 169$ e18 /2 = 84.5$ e18      84.5 e18/ 90e18 = 0.94


I really appreciate any help.

**Reply by EngrPips** [reply]
Hello, according to the protocol, you cannot deposit collateral worth $180 and mint DSC worth $100, the protocol doesn't work that way. If you deposit $180 into the protocol it will divide it into two, so you have $90 which is what your debt will be calculated against, and you can only mint up to half of the $90, which is $45.

**SIMPLE EXAMPLE**
COLLATERAL_VALUE = $180
COLLATERAL_VALUE_FOR_HEALTH_FATCOR_MATH = $90
MAXIMUM_AMOUNT_OF_DSC_TO_STAY_HEALTHY = $45

**Reply by miladad8** [reply]
The 180$ worth of deposit collateral is actually after the eth price crash. Basically this is the collateral value at the moment the user gets undercollateralized at the beginning part of the liquidation process. In `liquidate ` function when user gets undercollateral, first we calculate his healthFactor in order to compare it with healthFactor after the liquidator paid user's debt.

**Reply by EngrPips** [reply]
If that is the case then the calculation is correct , A user only gets unhealthy when their healthFactor is less than 1 and what is less than 1? Is it not 0?

**Reply by miladad8** [reply]
Well then why `startingUserHealthFactor` which is the healthFactor the moment user is getting under collateral calculated as 0.9? shouldn't be that 0 too?

**Reply by EngrPips** [reply]
It is either something is wrong in your `health_factor` function, or you minted the amount of DSC that is higher than what the user is supposed to mint to remain healthy.

**Comment by miladad8** [comment]
I sure do. thank you for your time :)

**Reply by EngrPips** [reply]
You are welcome.
